hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotCache.java

```diff
@@ -304,30 +305,28 @@ private void cleanupInternal() {
       OmSnapshot omSnapshot = (OmSnapshot) rcOmSnapshot.get();
       LOG.debug("Evicting OmSnapshot instance {} with table key {}",
           rcOmSnapshot, omSnapshot.getSnapshotTableKey());
-      // Sanity check
-      Preconditions.checkState(rcOmSnapshot.getTotalRefCount() == 0L,
-          "Illegal state: OmSnapshot reference count non-zero ("
-              + rcOmSnapshot.getTotalRefCount() + ") but shows up in the "
-              + "clean up list");
 
       final String key = omSnapshot.getSnapshotTableKey();
-      final ReferenceCounted<IOmMetadataReader, SnapshotCache> result =
-          dbMap.remove(key);
-      // Sanity check
-      Preconditions.checkState(rcOmSnapshot == result,
-          "Cache map entry removal failure. The cache is in an inconsistent "
-              + "state. Expected OmSnapshot instance: " + rcOmSnapshot
-              + ", actual: " + result + " for key: " + key);
-
-      pendingEvictionList.remove(result);
-
-      // Close the instance, which also closes its DB handle.
-      try {
-        ((OmSnapshot) rcOmSnapshot.get()).close();
-      } catch (IOException ex) {
-        throw new IllegalStateException("Error while closing snapshot DB", ex);
-      }
+      dbMap.computeIfPresent(key, (k, v) -> {
+        // Sanity check
+        Preconditions.checkState(rcOmSnapshot == v,
+            "Cache map entry removal failure. The cache is in an inconsistent "
+                + "state. Expected OmSnapshot instance: " + rcOmSnapshot
+                + ", actual: " + v + " for key: " + key);
+
+        // Close the instance, which also closes its DB handle.
+        if (rcOmSnapshot.getTotalRefCount() == 0L) {
+          try {
+            ((OmSnapshot) rcOmSnapshot.get()).close();
+          } catch (IOException ex) {
+            throw new IllegalStateException("Error while closing snapshot DB", ex);
+          }
+          return null;
+        }
+        return v;
+      });
 
+      pendingEvictionList.remove(rcOmSnapshot);
```

Comment:
I was thinking what if `rcOmSnapshot` is already removed from `pendingEvictionList` by another `get()` thread. It looks alright since it won't throw when the element doesn't exist in the list, only that the return value `boolean` would be `false`.

Should be good.

Reply:
Yeah it should be fine if it is removed already. Here we have to `remove()` the `rcOmSnapshot` instance from  `pendingEvictionList` regardless if we close the rocksdb or not. If it is still being referenced then we remove it from the `pendingEvictionList` , if it is not being referenced then we close() the DB instance and then `remove()`. Also the case that you mentioned if it is already removed then it will be a no-op.

Reply:
Could another `remove()` thread come in and append it back to the `pendingEvictionList`?

Reply:
`pendingEvictionList` is a synchronized set. If we are in the `cleanupInternal()` Then we already have the lock. I don't think another thread would be able to` remove()` in between this. I'm not sure I understand the append part during `remove()`, why do we need to append during `remove()`?

Reply:
Good point `cleanupInternal` is only called inside `synchronized (pendingEvictionList)`